backend/src/index.ts

```diff
@@ -0,0 +1,120 @@
+import { errorResponseBuilder, generateRandomPassword, isValidUser, parseUserForResponse } from './utils';
+
+const app = express();
+app.use(express.json());
+app.use(cors());
+
+app.post('/users/new', async (req: Request, res: Response) => {
+    const errorBuilder = errorResponseBuilder(res);
+
+    try {
+        const userData = req.body;
+
+        if (!isValidUser(req.body)) {
+            return errorBuilder.validationError();
+        }
+
+        const existingUserByEmail = await findUserByEmail(req.body.email);
+        if (existingUserByEmail) {
+            return errorBuilder.emailAlreadyInUse();
+        }
+
+        const existingUserByUsername = await findUserByUsername(req.body.username);
+        if (existingUserByUsername) {
+            return errorBuilder.usernameAlreadyTaken();
+        }
+
+        const user = await createUser({
+            data: { ...userData, password: generateRandomPassword(10) },
+        });
+
+        return res.status(201).json({
+            error: undefined,
+            data: parseUserForResponse(user),
+            succes: true,
+        });
+    } catch (error) {
+        return errorBuilder.serverError();
+    }
+});
+
+// Edit a user
+app.post('/users/edit/:userId', async (req: Request, res: Response) => {
+    const errorBuilder = errorResponseBuilder(res);
+
+    try {
+        const { id, password, ...userData } = req.body;
+
+        if (!isValidUser(userData)) {
+            return errorBuilder.validationError();
+        }
+
+        const foundUserById = await findUserById(Number(req.params.userId));
+
+        if (!foundUserById) {
+            return errorBuilder.userNotFound();
+        }
+
+        const foundUserByEmail = await findUserByEmail(userData.email);
+
+        // Allow passing the email unchanged
+        // Only throw error if we'd try to assing the same email to a different user
+        if (foundUserByEmail && foundUserByEmail.id !== foundUserById.id) {
+            return errorBuilder.emailAlreadyInUse();
+        }
+
+        const foundUserByUsername = await findUserByUsername(userData.username);
+
+        // Allow passing the username unchanged
+        // Only throw error if we'd try to assing the same username to a different user
+        if (foundUserByUsername && foundUserByUsername.id !== foundUserById.id) {
+            return errorBuilder.usernameAlreadyTaken();
+        }
+
+        const updatedUser = await updateUser(id, userData);
+
+        return res.status(200).json({
+            error: undefined,
+            data: parseUserForResponse(updatedUser),
+            succes: true,
+        });
+    } catch (error) {
+        return errorBuilder.serverError();
+    }
+});
+
+// Get a user by email
+app.get('/users', async (req: Request, res: Response) => {
+    const errorBuilder = errorResponseBuilder(res);
+
+    try {
+        const { email } = req.query;
+
+        if (!email) {
+            throw new Error();
```

Comment:
The specification doesn't specify anything in this case. But as we have some sort of data validation on other endpoints we can imagine returning a `400 InvalidData` error here. 

If not, as a user email is mandatory, you can also return a 404 here as  we know for sure the DB as 0 user without email.

Reply:
Thanks, in my 2nd PR I already modified this to throw a client error: https://github.com/enekesabel/DDDForum/pull/2/files#diff-c20200a666b149a045e365999f581db8e04687cfc9569bdf4c9477cab954d324L96-L98